### DISCUSSION TITLE
AbslStringify() on types you don't own

### DISCUSSION THREAD

**Discussion by jiawen** [abseil/abseil-cpp#1347] [General]
The new `AbslStringify()` API suggests that one should declare a `friend` function named `AbslStringify` to make it stringifyable. But is the `friend` qualifier necessary? It would be nice to stringify code that we don't own (e.g., from another library that doesn't want to take on Abseil as a dependency).

[The EnumWithStringify test](https://github.com/abseil/abseil-cpp/blob/afcb7de0cfe94ad2ec95dfd0380f74a4b972401c/absl/strings/substitute_test.cc#L258) suggests that the friend qualifier it not necessary. Is there any requirement for `friend` if the function were declared out of line and didn't need access to protected or private members?

**Comment by derekmauro** [comment]
This question is conflating two issues.

The first issue is defining `AbslStringify()` for types you don't own. We strongly urge you not to do this. What happens when two different libraries both decide to define  `AbslStringify()` for the same type? Those libraries become incompatible with each other. Therefore, to keep order, we strongly suggest that only the library that owns the type defines `AbslStringify()`. In the long run, if you want your library to play nice with other libraries, it is better to just add a normal `ToString()`-type function in your own library's namespace, and call that.

The second question is why is the `friend` qualifier necessary, instead of defining the function out-of-line, if the function doesn't need privileged access to the class? The answer here is that it isn't strictly necessary, but I think it is recommended because it is a tiny bit more efficient for the compiler when the compiler is performing overload resolution.

**Reply by jiawen** [reply]
Thanks for the fantastic explanation.  It's thoughtful design considerations like this that I love using Abseil.

Is there a good place to put this kind of design rationale for potential users? Some kind of FAQ or small update to the headers?

You probably don't want to update the blog or totw or anything.

(Obviously an Xoogler)
